src/tunnel.cc

```diff
@@ -922,7 +916,7 @@ TunnelStateData::tunnelEstablishmentDone(Http::TunnelerAnswer &answer)
 void
 TunnelStateData::notePeerReadyToShovel()
 {
-    server.resetCloseHandlerIfOpen(tunnelServerClosed, "tunnelServerClosed");
+    server.resetCloseHandler(tunnelServerClosed, "tunnelServerClosed");
```

Comment:
I do not see an 'explicit check' in this path: tunnelEstablishmentDone() -> notePeerReadyToShovel(). The connection should be open when connectedToPeer() is called ( and tunnelEstablishmentDone() is scheduled), but between this scheduling and calling the method it may close already.

Reply:
I agree that this code looks suspicious (at least), on several levels.

> The connection ... may close already.

If the connection is closed, why was not TunnelStateData close handler called? Did TunnelStateData have no handler when the connection started to close? Or do you mean that the connection is closing, and a TunnelStateData previous/current close handler call has been scheduled/queued already?

Moreover, the location of this reset call looks rather strange. Why are we resetting the close handler here (rather than earlier or later)? What state change does calling this method signify (that forces us to reset the handler)?

Reply:
> Or do you mean that the connection is closing, and a TunnelStateData previous/current close handler call has been scheduled/queued already?

Yes, this 'previous/current' is tlsServerClosed() handler, which is called on closures before the tunnel is fully established.

> Moreover, the location of this reset call looks rather strange.

Since tunnelServerClosed() handler is responsible for handling tunnel closures, I think we need to reset to tunnelServerClosed() handler when the tunnel is fully established. Also (according to the code) I think that it is fully established somewhere inside tunnelEstablishmentDone(), after we checked that the answer is possible. Assuming this, I reset the handler inside notePeerReadyToShovel() which is called in this case.

Reply:
> we need to reset to tunnelServerClosed() handler when the tunnel is fully established

I think that is one of the cases (for some definition of "fully established"). The other case is when the tunnel has failed in such a way that another attempt to establish a connection with the server is impossible or prohibited (for any reason). The common notion among these cases is that we are looking for a "point of no return".

I believe we reach that point of no return in two places:

1. Where connectDone() calls notePeerReadyToShovel() -- this handles to-origin use case.
2. Where connectedToPeer() discovers a _positive_ answer -- this handles to-peer use case. Note that this to-peer use cases also calls notePeerReadyToShovel(), but that happens later, well after the point of no return.

Placing closure handler reset in notePeerReadyToShovel() handles the first use case well, but it is the wrong spot because it causes problems for the second use case. The nature of those problems would vary from bugs to double-reset overheads, depending on how the rest of the code is changed, but those details are unimportant.

Branch commit f433d15 moves the close handler reset from notePeerReadyToShovel() to the two places enumerated above. Unfortunately, that change appears to expose more problems. Please see commit message TODO and commit XXXs.

Reply:
> Please see commit message TODO and commit XXXs.

I tried to address them at 3a9481f.

>  This solution requires determining whether the being-closed connection can be retried. If that determination is easy

I could not find how to determine this by means of available methods/flags so I had to add a new TunnelStateData::retriable flag for this purpose. There is a similar HappyConnOpener::retriable_ already.